Fix: Cannot generate unfiltered JSONL manifest in AnVIL (#7190)

nadove-ucsc opened the pull request:
<!--
This is the PR template for regular PRs against `develop`. Edit the URL in your
browser's location bar, appending either `&template=anvilprod-promotion.md`,
`&template=prod-promotion.md`, `&template=anvilprod-hotfix.md`, `&template=prod-
hotfix.md`, `&template=backport.md` or `&template=upgrade.md` to switch the
template.
-->

Connected issues: #7190


## Checklist


### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] On ZenHub, PR is connected to all issues it (partially) resolves
- [x] PR description links to connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] PR title references all connected issues
- [x] For each connected issue, there is at least one commit whose title references that issue

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title


### Author (partiality)

- [x] Added `p` tag to titles of partial commits
- [x] This PR is labeled `partial` <sub>or completely resolves all connected issues</sub>
- [x] This PR partially resolves each of the connected issues <sub>or does not have the `partial` label</sub>


### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or is not chained to another PR</sub>
- [x] The blocking PR is labeled `base` <sub>or this PR is not chained to another PR</sub>
- [x] This PR is labeled `chained` <sub>or is not chained to another PR</sub>


### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or the changes introduced by this PR will not require reindexing of any deployment</sub>
- [x] This PR is labeled `reindex:dev` <sub>or the changes introduced by it will not require reindexing of `dev`</sub>
- [x] This PR is labeled `reindex:anvildev` <sub>or the changes introduced by it will not require reindexing of `anvildev`</sub>
- [x] This PR is labeled `reindex:anvilprod` <sub>or the changes introduced by it will not require reindexing of `anvilprod`</sub>
- [x] This PR is labeled `reindex:prod` <sub>or the changes introduced by it will not require reindexing of `prod`</sub>
- [x] This PR is labeled `reindex:partial` and its description documents the specific reindexing procedure for `dev`, `anvildev`, `anvilprod` and `prod` <sub>or requires a full reindex or carries none of the labels `reindex:dev`, `reindex:anvildev`, `reindex:anvilprod` and `reindex:prod`</sub>
- [x] This PR and its connected issues are labeled `API` <sub>or this PR does not modify a REST API</sub>
- [x] Added `a` (`A`) tag to commit title for backwards (in)compatible changes <sub>or this PR does not modify a REST API</sub>
- [x] Updated REST API version number in `app.py` <sub>or this PR does not modify a REST API</sub>


### Author (upgrading deployments)

- [x] Ran `make docker_images.json` and committed the resulting changes <sub>or this PR does not modify `azul_docker_images`, or any other variables referenced in the definition of that variable</sub>
- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading deployments</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading deployments</sub>
- [x] This PR is labeled `upgrade` <sub>or does not require upgrading deployments</sub>
- [x] This PR is labeled `deploy:shared` <sub>or does not modify `docker_images.json`, and does not require deploying the `shared` component for any other reason</sub>
- [x] This PR is labeled `deploy:gitlab` <sub>or does not require deploying the `gitlab` component</sub>
- [x] This PR is labeled `deploy:runner` <sub>or does not require deploying the `runner` image</sub>


### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the none of the stable branches (`anvilprod` and `prod`) have temporary hotfixes for any of the issues connected to this PR</sub>


### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed fixups from prior reviews
- [x] Ran `make requirements_update` <sub>or this PR does not modify `requirements*.txt`, `common.mk`, `Makefile` and `Dockerfile`</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not modify `requirements*.txt`</sub>
- [x] This PR is labeled `reqs` <sub>or does not modify `requirements*.txt`</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not modify functionality that could affect the IT outcome</sub>


### Peer reviewer (after approval)

- [x] Actually approved the PR
- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] PR is awaiting requested review from system administrator
- [x] PR is assigned to only the system administrator


### System administrator (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] A comment to this PR details the completed security design review
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved connected issues to *Approved* column
- [x] PR is assigned to only the operator


### Operator (before pushing merge the commit)

- [x] Checked `reindex:…` labels and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Ran `_select dev.shared && CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select dev.gitlab && CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Ran `_select anvildev.shared && CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvildev.gitlab && CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Checked the items in the next section <sub>or this PR is labeled `deploy:gitlab`</sub>
- [x] PR is assigned to only the system administrator <sub>or this PR is not labeled `deploy:gitlab`</sub>


### System administrator

- [x] Background migrations for [`dev.gitlab`](https://gitlab.dev.singlecell.gi.ucsc.edu/admin/background_migrations) are complete <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Background migrations for [`anvildev.gitlab`](https://gitlab.anvil.gi.ucsc.edu/admin/background_migrations) are complete <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] PR is assigned to only the operator


### Operator (before pushing merge the commit)

- [x] Ran `_select dev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
- [x] Ran `_select anvildev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
- [x] Added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `dev`</sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `anvildev`</sub>
- [x] Started reindex in `sandbox` <sub>or this PR is not labeled `reindex:dev`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR is not labeled `reindex:anvildev`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR is not labeled `reindex:dev`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR is not labeled `reindex:anvildev`</sub>
- [x] All status checks passed and the PR is mergeable
- [x] The title of the merge commit starts with the title of this PR
- [x] Added PR # reference to merge commit title
- [x] Collected commit title tags in merge commit title <sub>but only included `p` if the PR is also labeled `partial`</sub>
- [x] Moved connected issues to *Merged lower* column in ZenHub
- [x] Moved blocked issues to *Triage* <sub>or no issues are blocked on the connected issues</sub>
- [x] Pushed merge commit to GitHub


### Operator (chain shortening)

- [x] Changed the target branch of the blocked PR to `develop` <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `chained` label from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the blocking relationship from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `base` label from this PR <sub>or this PR is not labeled `base`</sub>


### Operator (after pushing the merge commit)

- [x] Pushed merge commit to GitLab `dev`
- [x] Pushed merge commit to GitLab `anvildev`
- [x] Build passes on GitLab `dev`
- [x] Reviewed build logs for anomalies on GitLab `dev`
- [x] Build passes on GitLab `anvildev`
- [x] Reviewed build logs for anomalies on GitLab `anvildev`
- [x] Ran `_select dev.shared && make -C terraform/shared apply` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvildev.shared && make -C terraform/shared apply` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`


### Operator (reindex)

- [x] Deindexed all unreferenced catalogs in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Deindexed all unreferenced catalogs in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Deindexed specific sources in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Deindexed specific sources in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Indexed specific sources in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Indexed specific sources in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Checked for, triaged and possibly requeued messages in both fail queues in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Checked for, triaged and possibly requeued messages in both fail queues in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Emptied fail queues in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Emptied fail queues in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Restarted the Data Browser pipeline for the [ucsc/hca/dev branch](https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/data-browser/-/pipelines/new?ref=ucsc%2Fhca%2Fdev) on GitLab in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Restarted the Data Browser pipeline for the [ucsc/lungmap/dev branch](https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/data-browser/-/pipelines/new?ref=ucsc%2Flungmap%2Fdev) on GitLab in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Restarted `deploy_browser` job in the GitLab pipeline for this PR in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Restarted the Data Browser pipeline for the [ucsc/anvil/anvildev branch](https://gitlab.anvil.gi.ucsc.edu/ucsc/data-browser/-/pipelines/new?ref=ucsc%2Fanvil%2Fanvildev) on GitLab in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Restarted `deploy_browser` job in the GitLab pipeline for this PR in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>


### Operator

- [x] Propagated the `deploy:shared`, `deploy:gitlab`, `deploy:runner`, `API`, `reindex:partial`, `reindex:anvilprod` and `reindex:prod` labels to the next promotion PRs <sub>or this PR carries none of these labels</sub>
- [x] Propagated any specific instructions related to the `deploy:shared`, `deploy:gitlab`, `deploy:runner`, `API`, `reindex:partial`, `reindex:anvilprod` and `reindex:prod` labels, from the description of this PR to that of the next promotion PRs <sub>or this PR carries none of these labels</sub>
- [x] PR is assigned to no one


## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

nadove-ucsc committed:
Add type annotations

nadove-ucsc committed:
Refactor manifest paging

nadove-ucsc committed:
Apply type alias

nadove-ucsc committed:
Verbatim manifest generators use client side paging (#7190)

nadove-ucsc committed:
Extract superclass

nadove-ucsc committed:
Improve readability of manifest page uploading

nadove-ucsc committed:
Increase verbatim manifest page size for replica join (#7190)

nadove-ucsc committed:
Fix KeyError in storage service

nadove-ucsc committed:
Partition verbatim JSONL manifest by source (#7190)

nadove-ucsc committed:
Assert manifest partition size

nadove-ucsc committed:
Include orphans when filtering by source ID

dsotirho-ucsc reviewed (dismissed):
Approved.

hannes-ucsc commented on src/azul/service/storage_service.py:
```diff
@@ -329,7 +329,7 @@ def _handle_overwrite(self,
                           ):
         error = exception.response['Error']
-        code, condition = error['Code'], error['Condition']
+        code, condition = error['Code'], error.get('Condition')
```
PL, commit title misleading

nadove-ucsc replied:
Added comment to explain when the KeyError occurs. Agreed at PL that this is sufficient.

hannes-ucsc commented on src/azul/service/manifest_service.py:
```diff
@@ -460,6 +460,9 @@ def tuple_or_none(v):
     return v if v is None else tuple(v)
 
 
+type Cursor = tuple[str, str]
```
Needs documentation, e.g. what are the tuple elements. Reconcile with SortKey from elasticsearch_service.py. Double check the assumption that the first element is always a string.

hannes-ucsc commented on src/azul/service/manifest_service.py:
```diff
@@ -1224,7 +1224,41 @@ def storage(self):
+        return a, b
+
+
+class PagedManifestGenerator(PaginatingManifestGenerator):
```
"Paged" vs "Paginating" are synonymous so what's the difference?

nadove-ucsc replied:
"Paged" describes a result: the generator's output is a series of pages. "Paginating" describes a process: the generator uses client-side pagination when querying Elasticsearch. `PFBVerbatimManifestGenerator` is currently the only (non-abstract) `PaginatitingManifestGenerator` that is not a `PagedManifestGenerator`.

Note that `CompactManifestGenerator` and maybe others still don't use client-side paging. If they did, `PaginatingManifestGenerator` could be dissolved into the root `ManifestGenerator` ABC.

In the meantime I'll rename it to 'ClientSidePagingManigestGenerator` to try to make the name less confusing.

hannes-ucsc commented on src/azul/service/manifest_service.py:
```diff
@@ -2075,9 +2078,24 @@ class ReplicaKeys:
-        request = self._create_request()
-        for hit in request.scan():
+    def _get_all_hits(self,
+                      request_callback: Callable[[Cursor | None], Search]
```
Try to use an abstract method instead of the callback.

nadove-ucsc replied:
Both call sites, each using a different callback, occur within the same class, so I don't see how this is possible.

hannes-ucsc commented on src/azul/service/manifest_service.py:
```diff
@@ -1299,11 +1299,11 @@ def upload_part():
     page_size = 500
 
-    def _create_paged_request(self, partition: ManifestPartition) -> Search:
+    def _create_paged_request(self, search_after: Cursor | None) -> Search:
```
`Cursor` doesn't exist in this commit.

hannes-ucsc commented on src/azul/service/manifest_service.py:
```diff
@@ -1281,21 +1281,18 @@ def write(self,
                     if partition.is_last_page or buffer.tell() > self.part_size:
                         break
 
-                def upload_part():
```
PL please. The commit titles calls this a refactoring but this does not look like a refactoring to me.

nadove-ucsc replied:
Agreed at PL to change commit title to emphasize improvement in readability, without using the word "refactor".

hannes-ucsc commented on src/azul/service/manifest_service.py:
```diff
@@ -1220,7 +1223,41 @@ def storage(self):
+
+class PagedManifestGenerator(ClientSidePagingManifestGenerator):
     """
     A manifest generator whose output can be split over multiple concatenable
     IO streams.
```
```suggestion
    A manifest generator whose output is split over several concatenable
    segments, also known as pages.
```

And then describe how a page of ES results (as defined by the superclass) may or may not translate to a manifest page.

nadove-ucsc committed:
Add FIXME (#7290)

nadove-ucsc commented:
Ran indexing IT 5 times to check for nondeterminisms. Contents of unfiltered JSONL manifest is identical between this branch and `develop`, disregarding line order.

hannes-ucsc reviewed (changes requested):
```diff
Index: src/azul/service/manifest_service.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/service/manifest_service.py b/src/azul/service/manifest_service.py
--- a/src/azul/service/manifest_service.py	(revision 18a08a5ee38e3438d7d4d5e630955f29da40e3fd)
+++ b/src/azul/service/manifest_service.py	(date 1753149485623)
@@ -1262,11 +1262,15 @@
     A manifest generator whose output is split over several concatenable
     segments, also known as pages.
 
-    In some subclasses, e.g. `CompactManifestGenerator` and
-    `CurlManifestGenerator`, a manifest page corresponds to a page of hits from
-    a paginated Elasticsearch request. In others, e.g.
-    `JSONLVerbatimManifestGenerator`, the relationship between manifest pages
-    and ES pages is more complicated.
+    In some subclasses, e.g. CompactManifestGenerator and CurlManifestGenerator,
+    a manifest page corresponds to a page of hits from a paginated Elasticsearch
+    request. In others, e.g. `JSONLVerbatimManifestGenerator`, the relationship
+    between manifest pages and Elasticsearch pages is more complicated.
+
+    # REVIEW: I tend to use backquotes less often now, only if there is
+              ambiguity, as in "Python's `is` operator is useful", and then
+              since we aspire to use reStructureText in docstrings, this would
+              have to be double backquotes
     """
 
     @abstractmethod
@@ -1323,7 +1327,6 @@
                     assert buffer.tell() < 1.5 * 1024 ** 3
                     if partition.is_last_page or buffer.tell() > self.part_size:
                         break
-
                 if buffer.tell() > 0:
                     buffer.seek(0)
                     part_etag = self.storage.upload_multipart_part(buffer, partition.index + 1, upload)
@@ -2050,25 +2053,29 @@
 
     @property
     def include_orphans(self) -> bool:
+        
         # When filtering exclusively by properties of implicit hubs, e.g.,
         # data sets for AnVIL or projects for HCA, we include replicas of all
         # entities implicitly connected to the matching hubs, even replicas of
         # orphans, i.e., entities that aren't connected to files.
+        #
         plugin = self.metadata_plugin
         root_entity_fields = {
             field_name
             for field_name, field_path in plugin.field_mapping.items()
             if field_path[0] == 'contents' and field_path[1] == plugin.root_entity_type
         }
+        
         # For both HCA and AnVIL, these root entities are bijective with the
         # sources used for indexing, and filtering by a specific project
         # or dataset entity should produce the same results as filtering by
         # that entity's source.
         #
-        # The verbatim JSONL manifest generator temporarily inserts a source ID
-        # match into its provided filters in order to partition the manifest. If
-        # the source ID field were not included here, that insertion could
-        # cause orphans to be incorrectly dropped from the manifest.
+        # The verbatim JSONL generator temporarily inserts a source ID condition
+        # into its provided filters in order to partition the manifest. If the
+        # source ID field were not included below, that insertion would cause
+        # orphans to be absent from the manifest, which is incorrect.
+        #
         source_fields = {
             plugin.special_fields.source_id,
             plugin.special_fields.source_spec
@@ -2151,14 +2158,17 @@
             {'terms': {'entity_id.keyword': list(replica_ids)}}
         ]))
         request = request.extra(size=self.page_size)
-        # `_id` is currently the only replica field that is unique to each
+
+        # `_id` is currently the only index field that is unique to each replica
         # document (and thus results in an unambiguous total ordering). However,
-        # sorting by `_id` is unacceptably slow. To overcome the performance
-        # hit, we sort by a field that's *almost* unique to each replica, so
-        # that `_id` only needs to be loaded and compared in the infrequent
-        # event that it's needed as a tiebreaker.
+        # sorting just by `_id` is unacceptably slow, an Elasticsearch quirk. To
+        # overcome the performance hit, we sort by a field that's *almost*
+        # unique to each replica, so that `_id` only needs to be loaded and
+        # compared in the infrequent event that it's needed as a tiebreaker.
+        #
         # FIXME: ES DeprecationWarning for using _id as sort key
         #        https://github.com/DataBiosphere/azul/issues/7290
+        #
         request = request.sort('entity_id.keyword', '_id')
 
         def search(search_after: SortKey | None) -> Search:
@@ -2195,7 +2205,7 @@
 
         # It's safe to assume that the explicit sources are a subset of the
         # accessible sources. If they're not, an exception will be raised when
-        # the filters are applied.
+        # the filters are applied. REVIEW: "will be raised" or "would have been raised"?
         sources = self.filters.explicit.get(self.source_id_field,
                                             self.filters.source_ids)
         return sorted(sources)
@@ -2213,12 +2223,12 @@
             source_id = source_ids[partition.page_index]
             log.info('Listing replicas from source %r for manifest page %d',
                      source_id, partition.page_index)
-            source_partition_filters = self.filters.update({
+            partition_filter = {
                 self.source_id_field: {'is': [source_id]}
-            })
+            }
             original_filters = self.filters
             try:
-                self.filters = source_partition_filters
+                self.filters = original_filters.update(partition_filter)
                 replicas = self._list_replicas()
                 for replica in replicas:
                     entry = {
@@ -2229,7 +2239,10 @@
                     output.write('\n')
             finally:
                 self.filters = original_filters
-
+        # REVIEW: It feels odd that there is no `else` for the `if` above, or
+        #         that that `if` exists in the first place. You may need to
+        #         explain how the above `if` can be False when the statement
+        #         below seems to prevent that case.
         if partition.page_index + 1 < len(source_ids):
             return partition.next_page(file_name=None, search_after=None)
         else:
Index: test/azul_test_case.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/azul_test_case.py b/test/azul_test_case.py
--- a/test/azul_test_case.py	(revision 18a08a5ee38e3438d7d4d5e630955f29da40e3fd)
+++ b/test/azul_test_case.py	(date 1753145767697)
@@ -162,6 +162,7 @@
                 #        https://github.com/DataBiosphere/azul/issues/7290
                 RE('.*Loading the fielddata on the _id field is deprecated and will be removed in future versions.*'),
             },
+            # REVIEW: Out of curiosity, why do we need this twice, with two different classes?
             ElasticsearchWarning: {
                 # FIXME: ES DeprecationWarning for using _id as sort key
                 #        https://github.com/DataBiosphere/azul/issues/7290
```

nadove-ucsc committed:
Fix comment formatting

nadove-ucsc commented:
> REVIEW: "will be raised" or "would have been raised"?

"will be raised", because `self.source_ids()` is called _before_ the filters are applied. The assumption could be false at this point, but it's still **safe** because it is validated later, before an incorrect response would be produced.

>  REVIEW: Out of curiosity, why do we need this twice, with two different classes?

The `ElasticsearchWarning` is emitted directly by the ES library. The `DeprecationWarning` is [emitted by `DockerContainerTestCase`](https://github.com/DataBiosphere/azul/blob/81f8f2b1b3719c0d003f66dc6bc9135618551d1c/test/docker_container_test_case.py#L183).

The tests pass with no unexpected warnings with this patch applied:

```diff
diff --git a/test/azul_test_case.py b/test/azul_test_case.py
index 05941fb4b..378735dc3 100644
--- a/test/azul_test_case.py
+++ b/test/azul_test_case.py
@@ -157,10 +157,6 @@ class AzulTestCase(TestCase):
                 #        https://github.com/DataBiosphere/azul/issues/5953
                 'datetime.datetime.utcnow() is deprecated',
                 'datetime.datetime.utcfromtimestamp() is deprecated',
-
-                # FIXME: ES DeprecationWarning for using _id as sort key
-                #        https://github.com/DataBiosphere/azul/issues/7290
-                RE('.*Loading the fielddata on the _id field is deprecated and will be removed in future versions.*'),
             },
             ElasticsearchWarning: {
                 # FIXME: ES DeprecationWarning for using _id as sort key
diff --git a/test/docker_container_test_case.py b/test/docker_container_test_case.py
index dd92f638e..8995b1e23 100644
--- a/test/docker_container_test_case.py
+++ b/test/docker_container_test_case.py
@@ -177,9 +177,5 @@ class DockerContainerTestCase(AzulUnitTestCase):
 
     @classmethod
     def tearDownClass(cls):
-        for cached, containers in cls._containers.values():
-            for line in containers.logs().decode().split('\n'):
-                if 'deprecated' in line.lower():
-                    warnings.warn(line, DeprecationWarning)
         cls._kill_containers()
         super().tearDownClass()
```

hannes-ucsc commented on src/azul/service/manifest_service.py:
```diff
@@ -2124,21 +2187,61 @@ def file_name_extension(cls) -> str:
+
+        # It's safe to assume that the explicit sources are a subset of the
+        # accessible sources. If they're not, an exception will be raised when
+        # the filters are applied.
```
> > REVIEW: "will be raised" or "would have been raised"?

> "will be raised", because `self.source_ids()` is called _before_ the filters are applied. The assumption could be false at this point, but it's still **safe** because it is validated later, before an incorrect response would be produced.

I think I understand but I wouldn't word it that way. "It's safe to assume that it won't rain, because it it did, we have an umbrella" sounds confusing.

hannes-ucsc commented on src/azul/service/manifest_service.py:
```diff
@@ -2124,21 +2187,61 @@ def file_name_extension(cls) -> str:
+        # It's safe to assume that the explicit sources are a subset of the
+        # accessible sources. If they're not, an exception will be raised when
+        # the filters are applied.
+        sources = self.filters.explicit.get(self.source_id_field,
```
If `self.filters.explicit` contains an entry under the key `self.source_id_field`, then the value is something like `{"is": […]}` which I don't think is what you want.

nadove-ucsc replied:
You're right, apparently I overlooked testing this. I've now tested it with both accessible and inaccessible source filters, as well as without filters, and it behaves as expected.

hannes-ucsc replied:
<img width="1830" height="356" alt="CleanShot 2025-07-24 at 22 34 44@2x" src="https://github.com/user-attachments/assets/24b592e0-f5ba-46f4-8090-398e968feb5d" />


Can you please try covering this in the unit tests?

hannes-ucsc commented on test/azul_test_case.py:
```diff
@@ -154,6 +157,15 @@ class RE(str):
                 'datetime.datetime.utcnow() is deprecated',
                 'datetime.datetime.utcfromtimestamp() is deprecated',
+
+                # FIXME: ES DeprecationWarning for using _id as sort key
+                #        https://github.com/DataBiosphere/azul/issues/7290
+                RE('.*Loading the fielddata on the _id field is deprecated and will be removed in future versions.*'),
```
> >  REVIEW: Out of curiosity, why do we need this twice, with two different classes?
> 
> The `ElasticsearchWarning` is emitted directly by the ES library. The `DeprecationWarning` is [emitted by `DockerContainerTestCase`](https://github.com/DataBiosphere/azul/blob/81f8f2b1b3719c0d003f66dc6bc9135618551d1c/test/docker_container_test_case.py#L183).
> 
> The tests pass with no unexpected warnings with this patch applied:
> 
> ```diff
> diff --git a/test/azul_test_case.py b/test/azul_test_case.py
> index 05941fb4b..378735dc3 100644
> --- a/test/azul_test_case.py
> +++ b/test/azul_test_case.py
> @@ -157,10 +157,6 @@ class AzulTestCase(TestCase):
>                  #        https://github.com/DataBiosphere/azul/issues/5953
>                  'datetime.datetime.utcnow() is deprecated',
>                  'datetime.datetime.utcfromtimestamp() is deprecated',
> -
> -                # FIXME: ES DeprecationWarning for using _id as sort key
> -                #        https://github.com/DataBiosphere/azul/issues/7290
> -                RE('.*Loading the fielddata on the _id field is deprecated and will be removed in future versions.*'),
>              },
>              ElasticsearchWarning: {
>                  # FIXME: ES DeprecationWarning for using _id as sort key
> diff --git a/test/docker_container_test_case.py b/test/docker_container_test_case.py
> index dd92f638e..8995b1e23 100644
> --- a/test/docker_container_test_case.py
> +++ b/test/docker_container_test_case.py
> @@ -177,9 +177,5 @@ class DockerContainerTestCase(AzulUnitTestCase):
>  
>      @classmethod
>      def tearDownClass(cls):
> -        for cached, containers in cls._containers.values():
> -            for line in containers.logs().decode().split('\n'):
> -                if 'deprecated' in line.lower():
> -                    warnings.warn(line, DeprecationWarning)
>          cls._kill_containers()
>          super().tearDownClass()
> ```

Could you please provide an example of a matching line in the container output?

nadove-ucsc replied:
This is the only line that matches when running `service.test_manifest.TestManifests.test_verbatim_jsonl_manifest`:

```
{"type": "deprecation.elasticsearch", "timestamp": "2025-07-25T00:20:34,657Z", "level": "CRITICAL", "component": "o.e.d.i.m.IdFieldMapper", "cluster.name": "docker-cluster", "node.name": "ec3b2354ef02", "message": "Loading the fielddata on the _id field is deprecated and will be removed in future versions. If you require sorting or aggregating on this field you should also include the id in the body of your documents, and map this field as a keyword field that has [doc_values] enabled", "key": "id_field_data", "category": "mappings", "cluster.uuid": "He4JR8VZQra4rxiESeHTHA", "node.id": "BliL0MsuSG6A21XBtezknA"  }
```

hannes-ucsc replied:
Let's add your patch to the PR. If the deprecations are communicated via headers in the response, we don't need to propagate them through the back channel of the server logs. Also, with the caching of the ES container, we're reading the same logs several times, every time tearDownClass() is called.

hannes-ucsc reviewed (changes requested):
<img width="1688" height="226" alt="CleanShot 2025-07-24 at 22 38 07@2x" src="https://github.com/user-attachments/assets/c075fa1b-3d9d-4f3e-91f1-645e535e1a3c" />

It appears that we're only covering one page during the unit tests. Can we add coverage of more than one page?

https://github.com/DataBiosphere/azul/pull/7225#discussion_r2230194545

https://github.com/DataBiosphere/azul/pull/7225#discussion_r2230199261

nadove-ucsc committed:
Remove stale filters for permitted warnings (#2758, #2848, #2114)

nadove-ucsc committed:
Stop emitting deprecation warnings from container logs

If the deprecations are communicated via headers in the response, we don't need to propagate them through the back channel of the server logs.

hannes-ucsc reviewed (changes requested):
I suspect you might want to discuss some of these things in PL.

```diff
Index: test/service/test_manifest.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/service/test_manifest.py b/test/service/test_manifest.py
--- a/test/service/test_manifest.py	(revision a50751ea1a87c26a7840e0b683c21b9e99514da8)
+++ b/test/service/test_manifest.py	(date 1754453803807)
@@ -18,7 +18,10 @@
 from io import (
     BytesIO,
 )
-import itertools
+from itertools import (
+    combinations,
+    starmap,
+)
 import json
 import os
 from pathlib import (
@@ -1815,6 +1818,9 @@
 
 
 class TestVerbatimJSONLManifestSourcePartitioning(DCP1ManifestTestCase):
+
+    # REVIEW: Wouldn't three sources be sufficient?
+
     sources_by_bundle_uuid = {
         '3ac62c33-93e1-56b4-b857-59497f5d942d':
             DSSSourceRef(id='706cc417-9ed1-4c09-8341-0df38e374423',
@@ -1851,11 +1857,18 @@
                            for source in self.sources_by_bundle_uuid.values()
                        })
 
+    # REVIEW: The class and method name could be misconstrued to indicate that
+    #         the test covers partitioning of sources, which we also do but
+    #         somewhere else. This is testing the partitioning of manifests *by*
+    #         source.
+
     def test_source_partitioning(self):
 
         bundle_fqids = self.bundles()
 
-        # We can't assert every entity in the bundles because some HCA still
+        # REVIEW: There's a word missing in the comment below
+
+        # We can't assert every entity from the indexed bundles because some <WORD missing?> HCA still
         # lack replicas
         # FIXME: Some replicas are still missing for HCA
         #        https://github.com/DataBiosphere/azul/issues/6597
@@ -1869,19 +1882,22 @@
                 if filter_entity(ref.entity_type):
                     entity_ids_by_source_id[bundle_fqid.source.id].add(ref.entity_id)
 
+        # REVIEW: comment should explain what this does, or rather why this is needed
         assert all(
             len(entity_ids) >= 2
             for entity_ids in entity_ids_by_source_id.values()
         )
 
-        assert all(itertools.starmap(
+        # REVIEW: same here
+        assert all(starmap(
             set.isdisjoint,
-            itertools.combinations(entity_ids_by_source_id.values(), 2)
+            combinations(entity_ids_by_source_id.values(), 2)
         ))
 
-        for r in range(1, len(entity_ids_by_source_id) + 1):
-            for source_ids in itertools.combinations(entity_ids_by_source_id, r=r):
+        for num_sources in range(1, len(entity_ids_by_source_id) + 1):
+            for source_ids in combinations(entity_ids_by_source_id.keys(), num_sources):
                 with self.subTest(sources=source_ids):
+                    # REVIEW: Wouldn't it be more realistic to filter by project ID to test how that interacts with the implicit source filter?
                     response = self._get_manifest(ManifestFormat.verbatim_jsonl,
                                                   filters={'sourceId': {'is': list(source_ids)}})
                     manifest_rows = list(map(json.loads, response.content.decode().splitlines()))
@@ -1890,8 +1906,13 @@
                         for row in manifest_rows
                         if filter_entity(row['type'])
                     }
+                    # REVIEW: An explanatory comment is needed here as well …
                     for source_id, entity_ids in entity_ids_by_source_id.items():
                         expected_entity_ids = entity_ids if source_id in source_ids else set()
+                        # … especially the intersection with the actual value
+                        # smells fishy and needs justification, as the intersection
+                        # removes items which could to false positives when there
+                        # are a extra actual items.
                         self.assertEqual(expected_entity_ids, actual_entity_ids & entity_ids)
 
 
Index: src/azul/service/manifest_service.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/service/manifest_service.py b/src/azul/service/manifest_service.py
--- a/src/azul/service/manifest_service.py	(revision a50751ea1a87c26a7840e0b683c21b9e99514da8)
+++ b/src/azul/service/manifest_service.py	(date 1754440370803)
@@ -1835,12 +1835,24 @@
         hub_id: str
         replica_ids: list[str]
 
-    def _get_all_hits(self,
-                      request_callback: Callable[[SortKey | None], Search]
-                      ) -> Iterable[Hit]:
+    def _paginate_hits(self,
+                       request_factory: Callable[[SortKey | None], Search]
+                       ) -> Iterable[Hit]:
+        """
+        Yield all hits in every page of Elasticsearch hits in responses to
+        reauests that use client-side paging.
+
+        :param request_factory: A callable that returns a prepared Elasticsearch
+                                request for the given search-after key, with the
+                                appropriate filters and sorting applied. The
+                                returned request should yield one page worth of
+                                hits, starting at the first page (if the argument
+                                is None), or the hit right after the hit with
+                                given search-after key
+        """
         search_after = None
         while True:
-            request = request_callback(search_after)
+            request = request_factory(search_after)
             response = request.execute()
             if response.hits:
                 hit = None
@@ -1852,7 +1864,7 @@
                 break
 
     def _list_replica_keys(self) -> Iterable[ReplicaKeys]:
-        for hit in self._get_all_hits(self._create_paged_request):
+        for hit in self._paginate_hits(self._create_paged_request):
             document_ids = [
                 document_id
                 for entity_type in self.hot_entity_types
@@ -1910,10 +1922,10 @@
         #
         request = request.sort('entity_id.keyword', '_id')
 
-        def search(search_after: SortKey | None) -> Search:
+        def request_factory(search_after: SortKey | None) -> Search:
             return request.extra(search_after=search_after)
 
-        return self._get_all_hits(search)
+        return self._paginate_hits(request_factory)
 
 
 class JSONLVerbatimManifestGenerator(PagedManifestGenerator,
```

hannes-ucsc reviewed (changes requested):
```diff
Index: test/service/test_manifest.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/service/test_manifest.py b/test/service/test_manifest.py
--- a/test/service/test_manifest.py	(revision 72e89109d7319d4cc07771852582652fbe0b8cd7)
+++ b/test/service/test_manifest.py	(date 1754672677665)
@@ -1865,9 +1865,11 @@
     def test_manifest_partitioning_by_source(self):
 
         # We can't assert the presence of every entity from the indexed bundles
-        # because some HCA entities still lack replicas
+        # because some HCA entities still lack replicas.
+        #
         # FIXME: Some replicas are still missing for HCA
         #        https://github.com/DataBiosphere/azul/issues/6597
+        #
         def replicas_exist_for(entity_type: EntityType) -> bool:
             return entity_type in (
                 'project',
@@ -1893,6 +1895,7 @@
         # disjunctive. It's very easy to accidentally violate this invariant
         # while setting up this test, for example by choosing canned bundles
         # that came from the same source.
+        #
         assert all(starmap(
             set.isdisjoint,
             combinations(entity_ids_by_source_id.values(), 2)
@@ -1901,8 +1904,8 @@
         for num_sources in range(1, len(entity_ids_by_source_id) + 1):
             for source_ids in combinations(entity_ids_by_source_id, r=num_sources):
                 with self.subTest(sources=source_ids):
-                    response = self._get_manifest(ManifestFormat.verbatim_jsonl,
-                                                  filters={'sourceId': {'is': list(source_ids)}})
+                    filters = {'sourceId': {'is': list(source_ids)}}
+                    response = self._get_manifest(ManifestFormat.verbatim_jsonl, filters)
                     manifest_rows = list(map(json.loads, response.content.decode().splitlines()))
 
                     def entity_id(row: JSON) -> EntityID:
@@ -1930,13 +1933,12 @@
         accessible_source = list(self.sources_by_bundle_uuid.values())[0].id
         inaccessible_source = 'cafebabe-5b46-40e9-81c5-aaa7ebadf00d'
         with self.assertRaises(ForbiddenError) as e:
-            self._get_manifest(ManifestFormat.verbatim_jsonl,
-                               filters={'sourceId': {'is': [
-                                   accessible_source,
-                                   inaccessible_source
-                               ]}})
-        expected_args = ('Cannot filter by inaccessible sources',
-                         {inaccessible_source})
+            filters = {'sourceId': {'is': [accessible_source, inaccessible_source]}}
+            self._get_manifest(ManifestFormat.verbatim_jsonl, filters)
+        expected_args = (
+            'Cannot filter by inaccessible sources',
+            {inaccessible_source}
+        )
         self.assertEqual(expected_args, e.exception.args)
 
 
```

nadove-ucsc committed:
Add FIXME (#6597)

hannes-ucsc commented:
### Security design review

- [x] Security design review completed; this PR does **not** …
  - [x] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with Terra 
  - [x] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [x] … affect the permissions of external users like access to
    - TDR snapshots
  - [x] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [x] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [x] … affect monitoring of the system
  - [x] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [x] … add an interface that exposes sensitive or confidential data at the security boundary
  - [x] … affect the encryption of data at rest
  - [x] … require persistence of sensitive or confidential data that might require encryption at rest
  - [x] … require unencrypted transmission of data within the security boundary
  - [x] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [x] Documentation on any unchecked boxes is provided in comments below
